apm-agent-attach-cli/src/main/java/co/elastic/apm/attach/UserRegistry.java

```diff
@@ -222,5 +249,111 @@ public boolean isCurrentUser() {
+
+                long timeout = 5 * 1000L; // NOTE 5 second timeout!
+                long start = System.currentTimeMillis();
+                long now = start;
+                boolean isAlive = true;
+                byte[] buffer = new byte[64 * 1000];
+                try (InputStream in = spawnedProcess.getInputStream()) {
+                    // stop trying if the time elapsed exceeds the timeout
+                    while (isAlive && (now - start) < timeout) {
+                        while (in.available() > 0) {
+                            int lengthRead = in.read(buffer, 0, buffer.length);
+                            commandOutput.append(new String(buffer, 0, lengthRead));
+                        }
+                        try {
+                            Thread.sleep(50);
+                        } catch (InterruptedException e) {
+                            //no action, just means the next loop iteration is earlier
+                        }
+                        now = System.currentTimeMillis();
+                        // if it's not alive but there is still readable input, then continue reading
+                        isAlive = spawnedProcess.isAlive() || in.available() > 0;
+                    }
+                }
+                //would like to call waitFor(TIMEOUT) here, but that is 1.8+
+                //so will pause then have a go at getting the exit value
+                try {
+                    Thread.sleep(50);
+                } catch (InterruptedException e) {
+                    //no action, just means the next loop iteration is earlier
+                }
+                try {
+                    exitValue = spawnedProcess.exitValue();
+                } catch (IllegalThreadStateException e2) {
+                    //ignore and let the finally block have the last say
+                }
+
+                return new CommandOutput(commandOutput, exitValue, exception);
+            } catch (IOException e1) {
+                exception = e1;
+                return new CommandOutput(commandOutput, exitValue, exception);
+            } finally {
+                // Cleanup as well as we can
+                if (spawnedProcess != null && spawnedProcess.isAlive()) {
+                    spawnedProcess.destroy();
+                    try {
+                        Thread.sleep(50);
+                    } catch (InterruptedException e) {
+                        //no action, just means the next loop iteration is earlier
+                    }
+                    // when no longer need 1.7 compatibility, add these lines
+                    // try{Thread.sleep(50);}catch (InterruptedException e) {}
+                    // if (p.isAlive()) {
+                    // p.destroyForcibly();
+                    // }
+                }
+                try {
+                    exitValue = spawnedProcess.exitValue();
+                } catch (IllegalThreadStateException e2) {
+                    exception = e2;
```

Comment:
These `exitValue` and `exception` are ignored.
I believe you meat to do 
```
return new CommandOutput(commandOutput, exitValue, exception);
```
after the finally.

Reply:
hmm, I'll add comments to the file. Basically I call exitValue() in the main body just before returning from the body with the CommandOutput and if I don't do that then the exitValue() potentially doesn't get called until after the process exits in the finally. It turns out that somewhere in the JVM the process handling doesn't like it when I wait to the finally block, so sometimes it fails to correctly handle the spawned process.
The final call to exitValue() in the finally block is what finally sets the value in the variable, before the method exits and returns the CommandOutput (from either the main body or the catch clause). It's confusing because the JVM handling of spawned processes is confusing in 1.7 tbh. It gets better in 1.8
Please review after the next commit, if this doesn't make sense, or I got it wrong (entirely possible) then comment that, thx

Reply:
actually all of that doesn't matter, as I've found that the isAlive() call needs to be reimplmented for 1.7 compatibility so I have to call exitValue() earlier anyway. Will be in the next commit

Reply:
fixed the flow in commit 3a02a5cd094fd7eb4b033d3167e302463b90bf70 per your suggestion